source/libraries/CompileSolidity.ts

```diff
@@ -11,58 +11,66 @@ interface CompileContractsOutput {
-                let errors = "";
-                for (let error of compilerOutput.errors) {
-                    errors += error.formattedMessage + "\n";
-                }
-                throw new Error("The following errors/warnings were returned by solc:\n\n" + errors);
-            }
+            const result = fs.readFileSync(path, 'utf8');
+            return { contents: result };
+        } catch (error) {
+            return { error: error.message };
+        }
+    }
 
-            // Create output directory (if it doesn't exist)
-            mkdirp(this.contractOutputDirectoryPath, this.mkdirpCallback);
+    public async compileContracts(): Promise<CompileContractsOutput> {
+        // Check if all contracts are cached (and thus do not need to be compiled)
+        const stats = fs.statSync(this.contractOutputDirectoryPath + this.contractOutputFileName);
+        const lastCompiledTimestamp = stats.mtime;
+        const uncachedFiles = await recursiveReadDir(this.contractInputDirectoryPath, [function(file: string, stats: fs.Stats): boolean {return stats.isDirectory() || (stats.isFile() && path.extname(file) != ".sol") || (stats.isFile() && path.extname(file) == ".sol" && stats.mtime < lastCompiledTimestamp);}]);
+        if (uncachedFiles.length == 0) {
+            return { output: "Contracts in " + this.contractInputDirectoryPath + " have not been modified since last cache was created" };
+        }
 
-            // Output contract data to single file
-            const contractOutputFilePath = this.contractOutputDirectoryPath + "/" + this.contractOutputFileName;
-            let wstream: any = fs.createWriteStream(contractOutputFilePath);
-            for (let contract in compilerOutput.contracts) {
-                wstream.write(JSON.stringify(compilerOutput.contracts[contract]));
+        // Compile all contracts in the specified input directory
+        const compilerInputJson = await this.generateCompilerInput();
+        const compilerOutputJson: string = compileStandardWrapper(JSON.stringify(compilerInputJson), this.readCallback);
+        const compilerOutput: CompilerOutput = JSON.parse(compilerOutputJson);
+        if (compilerOutput.errors) {
+            let errors = "";
+            for (let error of compilerOutput.errors) {
+                errors += error.formattedMessage + "\n";
             }
-
-            return { output: "Contracts in " + this.contractInputDirectoryPath + " were successfully compiled by solc and saved to " + contractOutputFilePath};
-        } catch (error) {
-            throw error;
+            throw new Error("The following errors/warnings were returned by solc:\n\n" + errors);
         }
-    }
 
-    private ignoreFile(file: string, stats: fs.Stats): boolean {
-        return stats.isFile() && path.extname(file) != ".sol";
+        // Create output directory (if it doesn't exist)
+        mkdirp(this.contractOutputDirectoryPath, this.mkdirpCallback);
```

Comment:
Is there an async version of this function?

Reply:
Yeah, it looks like there's `async-mkdirp`.  I'll use that instead.